lofire-repo/src/commit.rs

```
@@ -0,0 +1,259 @@
+
+        // sign commit
+        let kp = match (author_privkey, author_pubkey) {
+            (PrivKey::Ed25519PrivKey(sk), PubKey::Ed25519PubKey(pk)) => [*sk, *pk].concat(),
+        };
+        let keypair = Keypair::from_bytes(kp.as_slice())?;
+        let sig_bytes = keypair.sign(content_ser.as_slice()).to_bytes();
+        let mut it = sig_bytes.chunks_exact(32);
+        let mut ss: Ed25519Sig = [[0; 32], [0; 32]];
+        ss[0].copy_from_slice(it.next().unwrap());
+        ss[1].copy_from_slice(it.next().unwrap());
+        let sig = Sig::Ed25519Sig(ss);
+        Ok(Commit::V0(CommitV0 { content, sig }))
+    }
+
+    /// Get commit signature
+    pub fn sig(&self) -> &Sig {
+        match self {
+            Commit::V0(c) => &c.sig,
+        }
+    }
+
+    /// Get commit content
+    pub fn content(&self) -> &CommitContentV0 {
+        match self {
+            Commit::V0(c) => &c.content,
+        }
+    }
+
+    /// Verify commit signature
+    pub fn verify_sig(&self) -> Result<(), SignatureError> {
+        let c = match self {
+            Commit::V0(c) => c,
+        };
+        let content_ser = serde_bare::to_vec(&c.content).unwrap();
+        let pubkey = match c.content.author {
+            PubKey::Ed25519PubKey(pk) => pk,
+        };
+        let pk = PublicKey::from_bytes(&pubkey)?;
+        let sig_bytes = match c.sig {
+            Sig::Ed25519Sig(ss) => [ss[0], ss[1]].concat(),
+        };
+        let sig = DalekSig::from_bytes(&sig_bytes)?;
+        pk.verify_strict(&content_ser, &sig)
+    }
+
+    /// Verify commit permissions
+    pub fn verify_perm(&self, body: &CommitBody, branch: &Branch) -> Result<(), CommitVerifyError> {
+        let content = self.content();
+        match branch.get_member(&content.author) {
+            Some(m) => {
+                if m.has_perm(body.to_type()) {
+                    return Ok(());
+                }
+            }
+            None => (),
```

i think it is also a permissionDenied if the author is not member of the branch

---

the None case should return an error